### PR TITLE
FAULT: Print ARM stack info on hard fault

### PR DESCRIPTION
**Pull request by henryhallam**
/cc @fnoble for review - this works as long as the stack isn't TOTALLY fucked.

### CODE REVIEW

File: src/error.c [hunk 1]
```diff
@@ -154,23 +154,37 @@ void NMIVector(void)
...
 void HardFaultVector(void)
 {
-  /* TODO: Stack trace / register dump per
-     http://www.freertos.org/Debugging-Hard-Faults-On-Cortex-M-Microcontrollers.html
-     https://blog.feabhas.com/2013/02/developing-a-generic-hard-fault-handler-for-arm-cortex-m3cortex-m4/
-     http://koti.kapsi.fi/jpa/stuff/other/stm32-hardfault-backtrace.html
-     TODO: ChibiOS thread status dump
+  /* TODO: ChibiOS thread status dump
      TODO: Use MSG_PANIC to avoid sprintf
      TODO: If stack is fucked, try to relocate it somewhere safe before pushing more things onto it.
```
Comment (fnoble) [inline src/error.c]: Does this already get done automatically?

File: src/error.c [hunk 2]
```diff
@@ -154,23 +154,37 @@ void NMIVector(void)
...
+     PM0214 section 2.3)
+  */
+  static uint32_t *psp, *lr;
+  asm("mrs %0, psp" : "=r"(psp) : : );
+  asm("mov %0, lr" : "=r"(lr) : : );
+  
   static char msg[256];
-  sprintf(msg, "HFSR=%08X CFSR=%08X MMFAR=%08X BFAR=%08X",
+  sprintf(msg, "HFSR=%08X CFSR=%08X MMFAR=%08X BFAR=%08X "
+          "PSP=%08X LR=%08X r0=%08X r1=%08X r2=%08X r3=%08X r12=%08X lr=%08X pc=%08X psr=%08X",
```
Comment (fnoble) [inline src/error.c]: Line length >>80

File: src/error.c [hunk 3]
```diff
@@ -154,23 +154,37 @@ void NMIVector(void)
...
+          (unsigned int)psp[0], (unsigned int)psp[1],
+          (unsigned int)psp[2], (unsigned int)psp[3],
+          (unsigned int)psp[4], (unsigned int)psp[5],
+          (unsigned int)psp[6], (unsigned int)psp[7]);
   _screaming_death(__func__, msg);
 };
 
 
 /** \} */
+//080041a0
```
Comment (fnoble) [inline src/error.c]: Remove this if it was just for testing?

**Comment by henryhallam** [comment]
Comments addressed

**Pull request merged**